akka-stream/src/main/scala/akka/stream/impl/AcknowledgePublisher.scala

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import akka.actor.Props
+import akka.stream.OverflowStrategy
+import akka.stream.OverflowStrategy._
+
+/**
+ * INTERNAL API
+ */
+private[akka] object AcknowledgePublisher {
+  def props(bufferSize: Int, overflowStrategy: OverflowStrategy) =
+    Props(new AcknowledgePublisher(bufferSize, overflowStrategy))
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] class AcknowledgePublisher(bufferSize: Int, overflowStrategy: OverflowStrategy)
+  extends ActorRefSourceActor(bufferSize, overflowStrategy) {
+
+  override def receiveElem: Receive = {
+    case elem if isActive ⇒
+      if (totalDemand > 0L) {
+        onNext(elem)
+        sendAck(true)
+      } else if (bufferSize == 0) {
+        log.debug("Dropping element because there is no downstream demand: [{}]", elem)
+        sendAck(false)
+      } else if (!buffer.isFull)
+        enqueueAndSendAck(elem)
+      else overflowStrategy match {
+        case DropHead ⇒
+          buffer.dropHead()
+          enqueueAndSendAck(elem)
+        case DropTail ⇒
+          buffer.dropTail()
+          enqueueAndSendAck(elem)
+        case DropBuffer ⇒
+          buffer.clear()
+          enqueueAndSendAck(elem)
+        case DropNew ⇒
+          sendAck(false)
+        case Fail ⇒
+          onErrorThenStop(new Fail.BufferOverflowException(s"Buffer overflow (max capacity was: $bufferSize)!"))
+        case Backpressure ⇒
+        // there is a precondition check in Source.actorRefSource factory method
```

Comment:
This Source would be much more useful if Backpressure mode would be supported. I.e. when you offer an element, and the buffer just got full by enqueueing that element, then the future should only complete after a new buffer space got freed. This way the Future would propagate properly the backpressure information.